compiler/rustc_query_system/src/query/plumbing.rs

```diff
@@ -360,7 +357,22 @@ where
     Qcx: QueryContext,
 {
     let state = Q::query_state(qcx);
-    match JobOwner::<'_, Q::Key, Qcx::DepKind>::try_start(&qcx, state, span, key) {
+    #[cfg(parallel_compiler)]
+    let state_lock = state.active.get_shard_by_value(&key).lock();
+    #[cfg(not(parallel_compiler))]
+    let state_lock = state.active.lock();
+
+    // For the parallel compiler we need to check both the query cache and query state structures
+    // while holding the state lock to ensure that 1) the query has not yet completed and 2) the
+    // query is not still executing.
+    if cfg!(parallel_compiler) && qcx.is_sync() {
+        if let Some((value, index)) = Q::query_cache(qcx).lookup(&key) {
+            qcx.dep_context().profiler().query_cache_hit(index.into());
+            return (value, Some(index));
+        }
+    }
```

Comment:
This check should happen in `JobOwner::try_start` which should return a `JobCompleted`.

Reply:
`try_start` isn't generic over query values, presumably for compile time reasons. Also `JobCompleted` is currently used as "job completed after blocking". It might make sense to refactor `try_execute_query` and `try_start`, but I'll wait for some PRs to merge before trying that.